File: kernel/src/ai/orchestration_manager.cpp
#include "jarvis/ai/orchestration_manager.h"
#include "jarvis/ai/agents_manager.h"
#include "jarvis/ai/models_manager.h"
#include "jarvis/ai/ollama_client.h"
#include <QSqlDatabase>
#include <QSqlQuery>
#include <QSqlError>
#include <QJsonDocument>
#include <QJsonObject>
#include <QJsonArray>
#include <QUuid>
#include <chrono>
#include <sstream>
#include <mutex>
#include <unordered_map>

namespace jarvis::ai {

class OrchestrationManager : public IOrchestrationManager {
public:
    OrchestrationManager(IModelsManager* models, IAgentsManager* agents,
                         const std::string& dbPath)
        : models_(models)
        , agents_(agents)
        , dbPath_(dbPath)
        , ollama_("http://localhost:11434")
    {
        initDb();
        loadConfig();
    }

    std::string executeQuery(const std::string& query) override {
        if (!config_.enabled || !streamCb_) {
            // Single agent mode
            return executeSingleAgent(query);
        }

        // Multi-agent mode
        auto queryId = QUuid::createUuid().toString(QUuid::Id128).toStdString();
        AgentTrace trace;
        trace.queryId = queryId;
        trace.query = query;

        emitChunk("\n");
        emitChunk("🧠 **Orquestrador**: Analisando sua pergunta...\n\n");

        // Step 1: Orchestrator decides which agents to use
        auto plan = orchestratorPlan(query);
        trace.orchestratorReasoning = plan.reasoning;

        emitChunk(plan.reasoning + "\n\n---\n\n");

        // Step 2: Execute specialist agents
        for (const auto& agentName : plan.agentNames) {
            auto agent = findAgent(agentName);
            if (agent.id.empty()) continue;

            emitChunk("**" + agentIcon(agent.specialty) + " " + agent.name + "**: trabalhando...\n\n");

            auto t1 = std::chrono::steady_clock::now();
            auto response = executeAgent(agent, query);
            auto t2 = std::chrono::steady_clock::now();

            AgentResult result;
            result.agentName = agent.name;
            result.specialty = agent.specialty;
            result.model = agent.model;
            result.response = response;
            result.tokensUsed = 0; // Would come from Ollama response
            result.latencyMs = std::chrono::duration_cast<std::chrono::milliseconds>(t2 - t1).count();
            trace.agentsConsulted.push_back(result);

            emitChunk(response + "\n\n---\n\n");
        }

        // Step 3: Critic review
        if (config_.criticEnabled) {
            emitChunk("🧠 **Critic Agent**: Revisando resposta...\n\n");
            auto review = criticReview(query, trace);
            trace.criticReview = review;
            emitChunk(review + "\n\n");
        }

        // Step 4: Consolidate final response
        auto finalResponse = consolidateResponse(query, trace);

        trace.finalResponse = finalResponse;
        {
            std::lock_guard<std::mutex> lock(mutex_);
            traces_[queryId] = trace;
        }

        return finalResponse;
    }

    AgentTrace getTrace(const std::string& queryId) override {
        std::lock_guard<std::mutex> lock(mutex_);
        auto it = traces_.find(queryId);
        if (it != traces_.end()) return it->second;
        return {};
    }

    OrchestrationConfig getConfig() override {
        return config_;
    }

    bool updateConfig(const OrchestrationConfig& config) override {
        config_ = config;
        QSqlQuery q(QSqlDatabase::database("ai-orchestration"));
        q.prepare(R"(
            UPDATE orchestration_config SET
                enabled = ?, orchestrator_model = ?, critic_enabled = ?,
                critic_temperature = ?, max_agents_per_query = ?,
                show_trace = ?, updated_at = ?
            WHERE id = 1
        )");
        q.addBindValue(config_.enabled ? 1 : 0);
        q.addBindValue(QString::fromStdString(config_.orchestratorModel));
        q.addBindValue(config_.criticEnabled ? 1 : 0);
        q.addBindValue(config_.criticTemperature);
        q.addBindValue(config_.maxAgentsPerQuery);
        q.addBindValue(config_.showTrace ? 1 : 0);
        q.addBindValue(QDateTime::currentDateTimeUtc().toString(Qt::ISODate));
        return q.exec();
    }

    void setStreamCallback(StreamCallback cb) override {
        streamCb_ = std::move(cb);
    }

private:
    IModelsManager* models_;
    IAgentsManager* agents_;
    std::string dbPath_;
    OllamaClient ollama_;
    OrchestrationConfig config_;
    StreamCallback streamCb_;
    std::mutex mutex_;
    std::unordered_map<std::string, AgentTrace> traces_;

    struct OrchestratorPlan {
        std::string reasoning;
        std::vector<std::string> agentNames;
    };

    void initDb() {
        QString connName = "ai-orchestration";
        QSqlDatabase db;
        if (QSqlDatabase::contains(connName)) {
            db = QSqlDatabase::database(connName);
        } else {
            db = QSqlDatabase::addDatabase("QSQLITE", connName);
        }
        db.setDatabaseName(QString::fromStdString(dbPath_));
        if (!db.isOpen()) db.open();
        QSqlQuery q(db);
        q.exec(R"(
            CREATE TABLE IF NOT EXISTS orchestration_config (
                id INTEGER PRIMARY KEY CHECK (id = 1),
                enabled INTEGER NOT NULL DEFAULT 1,
                orchestrator_model TEXT NOT NULL DEFAULT '',
                critic_enabled INTEGER NOT NULL DEFAULT 1,
                critic_temperature REAL NOT NULL DEFAULT 0.1,
                max_agents_per_query INTEGER NOT NULL DEFAULT 3,
                show_trace INTEGER NOT NULL DEFAULT 1,
                updated_at TEXT
            )
        )");
        q.exec("INSERT OR IGNORE INTO orchestration_config (id, enabled) VALUES (1, 1)");
    }

    void loadConfig() {
        QSqlQuery q(QSqlDatabase::database("ai-orchestration"));
        if (q.exec("SELECT * FROM orchestration_config WHERE id = 1") && q.next()) {
            config_.enabled = q.value("enabled").toBool();
            config_.orchestratorModel = q.value("orchestrator_model").toString().toStdString();
            config_.criticEnabled = q.value("critic_enabled").toBool();
            config_.criticTemperature = q.value("critic_temperature").toDouble();
            config_.maxAgentsPerQuery = q.value("max_agents_per_query").toInt();
            config_.showTrace = q.value("show_trace").toBool();
        }
    }

    std::string executeSingleAgent(const std::string& query) {
        auto agent = agents_->getDefaultAgent();
        if (agent.id.empty()) return "No agent configured.";

        return executeAgent(agent, query);
    }

    std::string executeAgent(const Agent& agent, const std::string& query) {
        OllamaGenerateRequest req;
        req.model = agent.model;
        req.prompt = query;
        req.system = agent.systemPrompt + "\n\nResponda em markdown.";
        req.temperature = agent.temperature;
        req.maxTokens = agent.maxTokens;
        req.stream = false;

        auto result = ollama_.generate(req);
        return result.response;
    }

    OrchestratorPlan orchestratorPlan(const std::string& query) {
        OrchestratorPlan plan;
        auto pool = agents_->getOrchestrationPool();

        if (pool.empty()) {
            plan.reasoning = "Nenhum agent especializado disponivel. Usando assistente padrao.";
            auto def = agents_->getDefaultAgent();
            if (!def.id.empty()) plan.agentNames.push_back(def.name);
            return plan;
        }

        // Build a prompt for the orchestrator to decide routing
        std::string agentList;
        for (size_t i = 0; i < pool.size() && i < 5; i++) {
            agentList += "- " + pool[i].name + " (" + pool[i].specialty + "): " +
                         pool[i].description + "\n";
        }

        std::string routingPrompt =
            "You are an orchestrator. Analyze this user query and decide which specialist "
            "agent(s) to route it to. Available agents:\n" + agentList +
            "\nUser query: " + query +
            "\n\nRespond with ONLY a comma-separated list of agent names that should handle this. "
            "If none match, respond with 'default'.";

        OllamaGenerateRequest req;
        req.model = config_.orchestratorModel.empty() ? pool[0].model : config_.orchestratorModel;
        req.prompt = routingPrompt;
        req.temperature = 0.1;
        req.maxTokens = 256;
        req.stream = false;

        try {
            auto result = ollama_.generate(req);
            auto response = result.response;

            // Parse comma-separated agent names
            std::stringstream ss(response);
            std::string name;
            while (std::getline(ss, name, ',')) {
                name.erase(0, name.find_first_not_of(" \n\r\t"));
                name.erase(name.find_last_not_of(" \n\r\t") + 1);
                if (!name.empty()) {
                    plan.agentNames.push_back(name);
                }
            }

            plan.reasoning = "📋 **Roteamento**: " + response;
        } catch (const std::exception& e) {
            plan.reasoning = "⚠️ Orchestrator error: " + std::string(e.what()) +
                             ". Usando agente padrao.";
            auto def = agents_->getDefaultAgent();
            if (!def.id.empty()) plan.agentNames.push_back(def.name);
        }

        if (plan.agentNames.empty()) {
            plan.agentNames.push_back("default");
        }

        return plan;
    }

    std::string criticReview(const std::string& query, const AgentTrace& trace) {
        // Build consolidated response for review
        std::string combined;
        for (const auto& r : trace.agentsConsulted) {
            combined += "Agent: " + r.agentName + "\n" + r.response + "\n\n";
        }

        std::string criticPrompt =
            "You are a critic agent. Review this response for:\n"
            "1. Does it fully answer the user's query?\n"
            "2. Is the information accurate?\n"
            "3. Is it well-formatted?\n"
            "4. Any improvements needed?\n\n"
            "User query: " + query + "\n\n"
            "Response to review:\n" + combined + "\n\n"
            "If approved, start with '✅ Aprovado'. If changes needed, start with '🔄 Refinar'.";

        OllamaGenerateRequest req;
        auto pool = agents_->getOrchestrationPool();
        req.model = pool.empty() ? "llama3.2:3b" : pool[0].model;
        req.prompt = criticPrompt;
        req.temperature = config_.criticTemperature;
        req.maxTokens = 512;
        req.stream = false;

        try {
            auto result = ollama_.generate(req);
            return result.response;
        } catch (...) {
            return "✅ Aprovado (critico indisponivel)";
        }
    }

    std::string consolidateResponse(const std::string& query, const AgentTrace& trace) {
        // For now, return the last agent's response or combine
        if (trace.agentsConsulted.empty()) return "Nenhuma resposta gerada.";
        return trace.agentsConsulted.back().response;
    }

    Agent findAgent(const std::string& name) {
        auto agents = agents_->listAgents();
        for (const auto& a : agents) {
            if (a.name == name || a.id == name) return a;
        }
        // Fallback to default
        return agents_->getDefaultAgent();
    }

    void emitChunk(const std::string& chunk) {
        if (streamCb_) streamCb_(chunk);
    }

    static std::string agentIcon(const std::string& specialty) {
        if (specialty == "code") return "\xf0\x9f\x92\xbb";
        if (specialty == "reasoning") return "\xf0\x9f\xa7\xa0";
        if (specialty == "embedding") return "\xf0\x9f\x93\x90";
        if (specialty == "chat") return "\xf0\x9f\x92\xac";
        return "\xf0\x9f\xa4\x96";
    }
};

IOrchestrationManager* createOrchestrationManager(
    IModelsManager* models, IAgentsManager* agents, const std::string& dbPath)
{
    return new OrchestrationManager(models, agents, dbPath);
}

} // namespace jarvis::ai

File: backend/jarvis/orchestration_manager.py
import datetime
import json
import time
import uuid
from dataclasses import dataclass, field
from typing import Callable

from jarvis.agents_manager import AgentsManager, Agent
from jarvis.database import Database
from jarvis.llm_gateway import LLMGateway, LLMProvider, LLMRequest, LLMMessage
from jarvis.models_manager import ModelsManager


@dataclass
class OrchestrationConfig:
    enabled: bool = True
    orchestrator_model: str = ""
    critic_enabled: bool = True
    critic_temperature: float = 0.1
    max_agents_per_query: int = 3
    show_trace: bool = True


@dataclass
class AgentResult:
    agent_name: str = ""
    specialty: str = ""
    model: str = ""
    response: str = ""
    tokens_used: int = 0
    latency_ms: int = 0


@dataclass
class AgentTrace:
    query_id: str = ""
    query: str = ""
    orchestrator_reasoning: str = ""
    agents_consulted: list[AgentResult] = field(default_factory=list)
    critic_review: str = ""
    final_response: str = ""


_SPECIALTY_ICONS = {
    "code": "💻",
    "reasoning": "🧠",
    "embedding": "📐",
    "chat": "💬",
}

_DEFAULT_CRITIC_PROMPT = (
    "You are a critic agent. Review this response for:\n"
    "1. Does it fully answer the user's query?\n"
    "2. Is the information accurate?\n"
    "3. Is it well-formatted?\n"
    "4. Any improvements needed?\n\n"
    "User query: {query}\n\n"
    "Response to review:\n{responses}\n\n"
    "If approved, start with '✅ Aprovado'. If changes needed, start with '🔄 Refinar'."
)

_DEFAULT_ROUTING_PROMPT = (
    "You are an orchestrator. Analyze this user query and decide which specialist "
    "agent(s) to route it to. Available agents:\n{agents}\n"
    "User query: {query}\n\n"
    "Respond with ONLY a comma-separated list of agent names that should handle this. "
    "If none match, respond with 'default'."
)


class OrchestrationManager:
    def __init__(
        self,
        models: ModelsManager,
        agents: AgentsManager,
        db: Database,
        llm_gateway: LLMGateway,
    ):
        self._models = models
        self._agents = agents
        self._db = db
        self._llm = llm_gateway
        self._stream_cb: Callable[[str], None] | None = None
        self._traces: dict[str, AgentTrace] = {}
        self._load_config()

    def set_stream_callback(self, cb: Callable[[str], None]) -> None:
        self._stream_cb = cb

    def execute_query(self, query: str) -> str:
        if not self._config.enabled or not self._stream_cb:
            return self._execute_single_agent(query)

        query_id = uuid.uuid4().hex
        trace = AgentTrace(query_id=query_id, query=query)

        self._emit_chunk("\n")
        self._emit_chunk("🧠 **Orquestrador**: Analisando sua pergunta...\n\n")

        plan_reasoning, agent_names = self._orchestrator_plan(query)
        trace.orchestrator_reasoning = plan_reasoning
        self._emit_chunk(plan_reasoning + "\n\n---\n\n")

        for agent_name in agent_names[:self._config.max_agents_per_query]:
            agent = self._find_agent(agent_name)
            if not agent:
                continue

            icon = _SPECIALTY_ICONS.get(agent.specialty, "🤖")
            self._emit_chunk(f"**{icon} {agent.name}**: trabalhando...\n\n")

            t1 = time.monotonic()
            response = self._execute_agent(agent, query)
            t2 = time.monotonic()

            result = AgentResult(
                agent_name=agent.name,
                specialty=agent.specialty,
                model=agent.model,
                response=response,
                latency_ms=int((t2 - t1) * 1000),
            )
            trace.agents_consulted.append(result)
            self._emit_chunk(response + "\n\n---\n\n")

        if self._config.critic_enabled:
            self._emit_chunk("🧠 **Critic Agent**: Revisando resposta...\n\n")
            review = self._critic_review(query, trace)
            trace.critic_review = review
            self._emit_chunk(review + "\n\n")

        final_response = self._consolidate_response(trace)
        trace.final_response = final_response
        self._traces[query_id] = trace

        return final_response

    def get_trace(self, query_id: str) -> AgentTrace | None:
        return self._traces.get(query_id)

    def get_config(self) -> OrchestrationConfig:
        return self._config

    def update_config(self, config: OrchestrationConfig) -> bool:
        self._config = config
        now = _now()
        self._db.execute(
            """
            UPDATE orchestration_config SET
                enabled = ?, orchestrator_model = ?, critic_enabled = ?,
                critic_temperature = ?, max_agents_per_query = ?,
                show_trace = ?, updated_at = ?
            WHERE id = 1
            """,
            (
                1 if config.enabled else 0,
                config.orchestrator_model,
                1 if config.critic_enabled else 0,
                config.critic_temperature,
                config.max_agents_per_query,
                1 if config.show_trace else 0,
                now,
            ),
        )
        return True

    def _load_config(self) -> None:
        row = self._db.fetchone(
            "SELECT * FROM orchestration_config WHERE id = 1"
        )
        if row:
            self._config = OrchestrationConfig(
                enabled=bool(row["enabled"]),
                orchestrator_model=row["orchestrator_model"],
                critic_enabled=bool(row["critic_enabled"]),
                critic_temperature=row["critic_temperature"],
                max_agents_per_query=row["max_agents_per_query"],
                show_trace=bool(row["show_trace"]),
            )
        else:
            self._config = OrchestrationConfig()

    def _execute_single_agent(self, query: str) -> str:
        agent = self._agents.get_default_agent()
        if not agent:
            return "No agent configured."
        return self._execute_agent(agent, query)

    def _execute_agent(self, agent: Agent, query: str) -> str:
        messages = [LLMMessage(role="user", content=query)]
        system = agent.system_prompt
        if system:
            system += "\n\nResponda em markdown."
        provider_str = agent.provider if agent.provider else "ollama"
        req = LLMRequest(
            provider=LLMProvider(provider_str),
            model=agent.model,
            messages=messages,
            system=system or "",
            temperature=agent.temperature,
            max_tokens=agent.max_tokens,
        )
        try:
            result = self._llm.generate(req)
            return result.content
        except Exception as e:
            return f"**Erro ao executar agente '{agent.name}':** {e}"

    def _orchestrator_plan(self, query: str) -> tuple[str, list[str]]:
        pool = self._agents.get_orchestration_pool()
        agent_names: list[str] = []

        if not pool:
            reasoning = (
                "Nenhum agent especializado disponivel. Usando assistente padrao."
            )
            default = self._agents.get_default_agent()
            if default:
                agent_names.append(default.name)
            return reasoning, agent_names

        agent_list = "\n".join(
            f"- {a.name} ({a.specialty}): {a.description}"
            for a in pool[:5]
        )

        routing_prompt = _DEFAULT_ROUTING_PROMPT.format(
            agents=agent_list, query=query
        )

        model = (
            self._config.orchestrator_model
            if self._config.orchestrator_model
            else pool[0].model
        )
        req = LLMRequest(
            model=model,
            messages=[LLMMessage(role="user", content=routing_prompt)],
            temperature=0.1,
            max_tokens=256,
        )

        try:
            result = self._llm.generate(req)
            response = result.content.strip()
            agent_names = [
                name.strip()
                for name in response.split(",")
                if name.strip()
            ]
            reasoning = f"📋 **Roteamento**: {response}"
        except Exception as e:
            reasoning = (
                f"⚠️ Orchestrator error: {e}. Usando agente padrao."
            )
            default = self._agents.get_default_agent()
            if default:
                agent_names.append(default.name)

        if not agent_names:
            agent_names.append("default")

        return reasoning, agent_names

    def _critic_review(self, query: str, trace: AgentTrace) -> str:
        combined = "\n\n".join(
            f"Agent: {r.agent_name}\n{r.response}"
            for r in trace.agents_consulted
        )
        prompt = _DEFAULT_CRITIC_PROMPT.format(
            query=query, responses=combined
        )

        pool = self._agents.get_orchestration_pool()
        model = pool[0].model if pool else "llama3.2"

        req = LLMRequest(
            model=model,
            messages=[LLMMessage(role="user", content=prompt)],
            temperature=self._config.critic_temperature,
            max_tokens=512,
        )
        try:
            result = self._llm.generate(req)
            return result.content
        except Exception:
            return "✅ Aprovado (critico indisponivel)"

    def _consolidate_response(self, trace: AgentTrace) -> str:
        if not trace.agents_consulted:
            return "Nenhuma resposta gerada."
        return trace.agents_consulted[-1].response

    def _find_agent(self, name: str) -> Agent | None:
        agents = self._agents.list_agents()
        for a in agents:
            if a.name == name or a.id == name:
                return a
        return self._agents.get_default_agent()

    def _emit_chunk(self, chunk: str) -> None:
        if self._stream_cb:
            self._stream_cb(chunk)


def _now() -> str:
    return datetime.datetime.now(datetime.timezone.utc).strftime(
        "%Y-%m-%dT%H:%M:%S.%fZ"
    )
